parlai/utils/fsdp.py

```diff
@@ -7,22 +7,46 @@
 """
 Utility functions for FullyShardedDataParallel.
 """
-
 import contextlib
+import functools
+import torch
+import torch.distributed
+from torch.distributed.algorithms.join import Join, Joinable, JoinHook
 import torch.nn
+
+from parlai.scripts.train_model import TrainLoop
 from parlai.utils.distributed import is_distributed, get_dist_group
 
 try:
-    from fairscale.nn.wrap.auto_wrap import wrap
-    from fairscale.nn.wrap.auto_wrap import enable_wrap as fairscale_enable_wrap
-    from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
+    import torch
+    import torch.distributed
+    import torch.distributed.fsdp
+    from torch.distributed.fsdp.wrap import (
+        wrap,
+        enable_wrap,
+        transformer_auto_wrap_policy,
+    )
+    from torch.distributed.fsdp.fully_sharded_data_parallel import (
+        FullyShardedDataParallel as FSDP,
+        ShardingStrategy,
+        MixedPrecision,
+        BackwardPrefetch,
+    )
 
+    PYTORCH_FSDP_AVAILABLE = True
     FSDP_AVAILABLE = True
 except ImportError:
-    FSDP_AVAILABLE = False
+    PYTORCH_FSDP_AVAILABLE = False
+    try:
+        from fairscale.nn.wrap.auto_wrap import wrap, enable_wrap
+        from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
```

Comment:
why keep around fairscale support?

Reply:
so getting rid of fairscale would force anyone who wants to use distributed training to be on pytorch >=1.12. Is that a reasonable ask, you think?

Reply:
That's your call but pytorch 1.13 is out (and I'm using it successfully)

Reply:
i've removed fairscale FSDP